File: google-drive/typescript/src/services/changes.ts
/**
 * Changes service for Google Drive API.
 *
 * Monitors and tracks changes to files and folders.
 */

import {
  Change,
  ChangeList,
  StartPageToken,
  Channel,
  GetStartPageTokenParams,
  ListChangesParams,
  WatchChangesRequest,
} from '../types';

/**
 * HTTP transport interface.
 */
export interface ChangeTransport {
  request<T>(url: string, options: RequestOptions): Promise<T>;
}

/**
 * Request options.
 */
export interface RequestOptions {
  method?: string;
  body?: any;
  params?: Record<string, any>;
}

/**
 * Changes service interface.
 */
export interface ChangesService {
  /**
   * Get the start page token for change tracking.
   *
   * @param params - Optional parameters
   * @returns The start page token
   */
  getStartPageToken(params?: GetStartPageTokenParams): Promise<StartPageToken>;

  /**
   * List changes since a page token.
   *
   * @param pageToken - Start page token
   * @param params - Optional list parameters
   * @returns Change list with pagination
   */
  list(pageToken: string, params?: ListChangesParams): Promise<ChangeList>;

  /**
   * List all changes with auto-pagination.
   *
   * @param startPageToken - Start page token
   * @param params - Optional list parameters
   * @returns Async iterable of changes
   */
  listAll(startPageToken: string, params?: ListChangesParams): AsyncIterable<Change>;

  /**
   * Watch for changes via push notifications.
   *
   * @param pageToken - Start page token
   * @param request - Watch request
   * @returns The notification channel
   */
  watch(pageToken: string, request: WatchChangesRequest): Promise<Channel>;

  /**
   * Stop watching for changes.
   *
   * @param channel - The channel to stop
   */
  stopWatch(channel: Channel): Promise<void>;
}

/**
 * Implementation of ChangesService.
 */
export class ChangesServiceImpl implements ChangesService {
  private readonly baseUrl = 'https://www.googleapis.com/drive/v3';

  constructor(private transport: ChangeTransport) {}

  async getStartPageToken(params?: GetStartPageTokenParams): Promise<StartPageToken> {
    return this.transport.request<StartPageToken>(
      `${this.baseUrl}/changes/startPageToken`,
      {
        method: 'GET',
        params: params as any,
      }
    );
  }

  async list(pageToken: string, params?: ListChangesParams): Promise<ChangeList> {
    return this.transport.request<ChangeList>(`${this.baseUrl}/changes`, {
      method: 'GET',
      params: {
        pageToken,
        ...params,
      } as any,
    });
  }

  async *listAll(startPageToken: string, params?: ListChangesParams): AsyncIterable<Change> {
    let pageToken: string | undefined = startPageToken;

    while (pageToken) {
      const response = await this.list(pageToken, params);

      for (const change of response.changes) {
        yield change;
      }

      // If newStartPageToken is present, we've reached the end
      if (response.newStartPageToken) {
        break;
      }

      pageToken = response.nextPageToken;
    }
  }

  async watch(pageToken: string, request: WatchChangesRequest): Promise<Channel> {
    return this.transport.request<Channel>(`${this.baseUrl}/changes/watch`, {
      method: 'POST',
      params: { pageToken },
      body: request,
    });
  }

  async stopWatch(channel: Channel): Promise<void> {
    await this.transport.request<void>(`${this.baseUrl}/channels/stop`, {
      method: 'POST',
      body: {
        id: channel.id,
        resourceId: channel.resourceId,
      },
    });
  }
}

/**
 * Mock implementation for testing.
 */
export class MockChangesService implements ChangesService {
  private changes: Change[] = [];
  private pageToken = 'start-token';
  private channels = new Map<string, Channel>();

  async getStartPageToken(params?: GetStartPageTokenParams): Promise<StartPageToken> {
    return {
      kind: 'drive#startPageToken',
      startPageToken: this.pageToken,
    };
  }

  async list(pageToken: string, params?: ListChangesParams): Promise<ChangeList> {
    return {
      kind: 'drive#changeList',
      newStartPageToken: 'new-start-token',
      changes: this.changes,
    };
  }

  async *listAll(startPageToken: string, params?: ListChangesParams): AsyncIterable<Change> {
    const response = await this.list(startPageToken, params);
    for (const change of response.changes) {
      yield change;
    }
  }

  async watch(pageToken: string, request: WatchChangesRequest): Promise<Channel> {
    const channel: Channel = {
      kind: 'api#channel',
      id: request.id,
      resourceId: `resource-${Date.now()}`,
      resourceUri: `https://www.googleapis.com/drive/v3/changes?pageToken=${pageToken}`,
      expiration: request.expiration,
      token: request.token,
      address: request.address,
      type: request.type,
      params: request.params,
    };

    this.channels.set(channel.id, channel);
    return channel;
  }

  async stopWatch(channel: Channel): Promise<void> {
    this.channels.delete(channel.id);
  }

  /**
   * Helper method to add a mock change.
   */
  addChange(change: Change): void {
    this.changes.push(change);
  }

  /**
   * Helper method to clear all changes.
   */
  clearChanges(): void {
    this.changes = [];
  }
}

/**
 * Create a mock changes service.
 */
export function createMockChangesService(): MockChangesService {
  return new MockChangesService();
}

File: google-drive/rust/src/services/changes.rs
//! Change tracking service.

use crate::client::GoogleDriveClient;
use crate::error::GoogleDriveResult;
use crate::types::*;
use futures::stream::Stream;
use serde::{Deserialize, Serialize};

/// Service for change tracking.
pub struct ChangesService<'a> {
    client: &'a GoogleDriveClient,
}

impl<'a> ChangesService<'a> {
    /// Creates a new changes service.
    pub fn new(client: &'a GoogleDriveClient) -> Self {
        Self { client }
    }

    /// Gets the start page token for change tracking.
    pub async fn get_start_page_token(
        &self,
        params: Option<GetStartPageTokenParams>,
    ) -> GoogleDriveResult<StartPageToken> {
        if let Some(p) = params {
            self.client
                .get_with_params("/changes/startPageToken", &p)
                .await
        } else {
            self.client.get("/changes/startPageToken").await
        }
    }

    /// Lists changes since a page token.
    pub async fn list(
        &self,
        page_token: &str,
        params: Option<ListChangesParams>,
    ) -> GoogleDriveResult<ChangeList> {
        let mut list_params = params.unwrap_or_default();
        list_params.page_token = Some(page_token.to_string());
        self.client.get_with_params("/changes", &list_params).await
    }

    /// Lists all changes with auto-pagination.
    pub fn list_all(
        &self,
        start_page_token: &str,
        params: Option<ListChangesParams>,
    ) -> impl Stream<Item = GoogleDriveResult<Change>> + '_ {
        let initial_token = start_page_token.to_string();
        async_stream::stream! {
            let mut page_token = Some(initial_token);

            loop {
                if let Some(token) = page_token {
                    let result = self.list(&token, params.clone()).await;

                    match result {
                        Ok(change_list) => {
                            for change in change_list.changes {
                                yield Ok(change);
                            }

                            // Check if there are more pages
                            if let Some(next) = change_list.next_page_token {
                                page_token = Some(next);
                            } else {
                                break;
                            }
                        }
                        Err(e) => {
                            yield Err(e);
                            break;
                        }
                    }
                } else {
                    break;
                }
            }
        }
    }

    /// Watches for changes via push notifications.
    pub async fn watch(
        &self,
        page_token: &str,
        request: WatchChangesRequest,
    ) -> GoogleDriveResult<Channel> {
        let path = format!("/changes/watch?pageToken={}", page_token);
        self.client.post(&path, &request).await
    }

    /// Stops watching for changes.
    pub async fn stop_watch(&self, channel: &Channel) -> GoogleDriveResult<()> {
        let request = StopChannelRequest {
            id: channel.id.clone(),
            resource_id: channel.resource_id.clone(),
        };
        self.client.post_no_response("/channels/stop", &request).await
    }
}

// Request/Response types

/// Parameters for getting start page token.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct GetStartPageTokenParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub drive_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_all_drives: Option<bool>,
}

/// Parameters for listing changes.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ListChangesParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub drive_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_corpus_removals: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_items_from_all_drives: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_removed: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_size: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub restrict_to_my_drive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub spaces: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_all_drives: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<String>,
}

/// Request to watch for changes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WatchChangesRequest {
    pub id: String,
    #[serde(rename = "type")]
    pub channel_type: String,
    pub address: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expiration: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
}

/// Request to stop a channel.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct StopChannelRequest {
    id: String,
    resource_id: String,
}
